From ff8fd50bd2dd1a49cfac0fa02bd34bac5ce987d2 Mon Sep 17 00:00:00 2001
From: Amit Shah <amit.shah@redhat.com>
Date: Tue, 12 Aug 2014 11:30:52 -0400
Subject: [char] hwrng: cleanup in hwrng_register()

Message-id: <56c939cd7a9292d3e577d571fc3781ee7c24140c.1407842211.git.amit.shah@redhat.com>
Patchwork-id: 87649
O-Subject: [RHEL7.1 kernel PATCH 2/9] hwrng: cleanup in hwrng_register()
Bugzilla: 1119299
RH-Acked-by: Amos Kong <akong@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Herbert Xu <herbert.xu@redhat.com>

From: Dan Carpenter <dan.carpenter@oracle.com>

My static checker complains that:

 drivers/char/hw_random/core.c:341 hwrng_register()
 warn: we tested 'old_rng' before and it was 'false'

The problem is that sometimes we test "if (!old_rng)" and sometimes we
test "if (must_register_misc)".  The static checker knows they are
equivalent but a human being reading the code could easily be confused.

I have simplified the code by removing the "must_register_misc" variable
and I have removed the redundant check on "if (!old_rng)".

Signed-off-by: Dan Carpenter <dan.carpenter@oracle.com>
Reviewed-by: Rusty Russell <rusty@rustcorp.com.au>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
(cherry picked from commit d167b6e1fb8ad386b17485ca88804d14f1695805)
Signed-off-by: Amit Shah <amit.shah@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/char/hw_random/core.c b/drivers/char/hw_random/core.c
index a0f7724..cf49f1c 100644
--- a/drivers/char/hw_random/core.c
+++ b/drivers/char/hw_random/core.c
@@ -302,7 +302,6 @@ err_misc_dereg:
 
 int hwrng_register(struct hwrng *rng)
 {
- int must_register_misc;
  int err = -EINVAL;
  struct hwrng *old_rng, *tmp;
 
@@ -327,7 +326,6 @@ int hwrng_register(struct hwrng *rng)
    goto out_unlock;
  }
 
- must_register_misc = (current_rng == NULL);
  old_rng = current_rng;
  if (!old_rng) {
   err = hwrng_init(rng);
@@ -336,13 +334,11 @@ int hwrng_register(struct hwrng *rng)
   current_rng = rng;
  }
  err = 0;
- if (must_register_misc) {
+ if (!old_rng) {
   err = register_miscdev();
   if (err) {
-   if (!old_rng) {
-    hwrng_cleanup(rng);
-    current_rng = NULL;
-   }
+   hwrng_cleanup(rng);
+   current_rng = NULL;
    goto out_unlock;
   }
  }
-- 
1.7.1